designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,499 @@
+import example from "eslint-plugin-example";
+
+export default [
+    {
+        ...js.configs.recommended
+        files: ["**/src/*.js"],
+    },
+    {
+        ...example.configs.recommended
+        files: ["**/src/*.js"],
+    },
+];
+```
+
+If the objects in `extends` contain `files` or `ignores`, then ESLint will merge those values with the values found in the config using `extends`. For example:
+
+```js
+import globals from "globals";
+
+const config1 = {
+    name: "config1",
+    files: ["**/*.cjs"],
+    languageOptions: {
+        sourceType: "commonjs",
+        globals: {
+            ...globals.node
+        }
+    }
+};
+
+const config2 = {
+    name: "config2",
+    files: ["**/*.js"],
+    ignores: ["**/tests"],
+    rules: {
+        "no-console": "error"
+    }
+};
+
+
+export default [
+
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        extends: [config1, config2],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Here, the `files` keys will be combined and the `ignores` key will be inherited, resulting in a final config that looks like this:
+
+```js
+
+export default [
+
+    // combined files
+    {
+        name: "myconfig > config1",
+        files: ["**/src/*.js", "**/*.cjs"],
+        languageOptions: {
+            sourceType: "commonjs",
+            globals: {
+                ...globals.node
+            }
+        }
+    },
+
+    // combined files and inherited ignores
+    {
+        name: "myconfig > config2",
+        files: ["**/src/*.js", "**/*.js"],
+        ignores: ["**/tests"],
+        rules: {
+            "no-console": "error"
+        }
+    },
+
+    // original config
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Note that the `name` key is generated for calculated config objects so that it indicates the inheritance from another config that doesn't exist in the final representation of the config array.
+
+#### Extending Arrays
+
+Arrays can also be used in `extends` (to eliminate the guesswork of what type a config is). When evaluating `extends`, ESLint internally calls `.flat()` on the array and then processes the config objects as discussed in the previous example. Consider the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [
+            [js.configs.recommended, example.configs.recommended]
+        ]
+    }
+
+];
+```
+
+This is equivalent to the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+The extended objects are evaluated in the same order and result in the same final config.
+
+#### Extending Named Configs
```

Comment:
Noted.

Reply:
I think supporting nested arrays is good, as currently most recommended configs are non-arrays and if they would have to become arrays then that would become a breaking change unless it can be consumed in the same way as a non-array